contracts/icp/proxy-contract/src/mutate.rs

```diff
@@ -0,0 +1,304 @@
+        return Err("request expired".to_string());
+    }
+
+    // Check membership
+    if !check_member(&request.signer_id).await? {
+        return Err("signer is not a member".to_string());
+    }
+
+    match &request.kind {
+        ICRequestKind::Propose { proposal } => {
+            let num_proposals = PROXY_CONTRACT.with(|contract| {
+                let contract = contract.borrow();
+                contract
+                    .num_proposals_pk
+                    .get(&proposal.author_id)
+                    .copied()
+                    .unwrap_or(0)
+            });
+
+            internal_create_proposal(proposal.clone(), num_proposals)
+        }
+        ICRequestKind::Approve { approval } => {
+            internal_approve_proposal(
+                approval.signer_id.clone(),
+                approval.proposal_id.clone(),
+                approval.added_timestamp,
+            )
+            .await
+        }
+    }
+}
+
+async fn internal_approve_proposal(
+    signer_id: ICSignerId,
+    proposal_id: ICProposalId,
+    _added_timestamp: u64,
+) -> Result<Option<ICProposalWithApprovals>, String> {
+    // First phase: Update approvals and check if we need to execute
+    let should_execute = PROXY_CONTRACT.with(|contract| {
+        let mut contract = contract.borrow_mut();
+
+        // Check if proposal exists
+        if !contract.proposals.contains_key(&proposal_id) {
+            return Err("proposal does not exist".to_string());
+        }
+
+        let approvals = contract.approvals.entry(proposal_id.clone()).or_default();
+
+        if approvals.contains(&signer_id) {
+            return Err("proposal already approved".to_string());
+        }
+
+        approvals.insert(signer_id);
+
+        Ok(approvals.len() as u32 >= contract.num_approvals)
+    })?;
+
+    // Execute if needed
+    if should_execute {
+        execute_proposal(&proposal_id).await?;
+    }
+
+    // Build final response
+    PROXY_CONTRACT.with(|contract| {
+        let contract = contract.borrow();
+        build_proposal_response(&*contract, proposal_id)
+    })
+}
+
+async fn execute_proposal(proposal_id: &ICProposalId) -> Result<(), String> {
+    let proposal = PROXY_CONTRACT.with(|contract| {
+        let contract = contract.borrow();
+        contract
+            .proposals
+            .get(proposal_id)
+            .cloned()
+            .ok_or_else(|| "proposal does not exist".to_string())
+    })?;
+
+    // Execute each action
+    for action in proposal.actions {
+        match action {
+            ICProposalAction::ExternalFunctionCall {
+                receiver_id,
+                method_name,
+                args,
+                deposit: _,
+            } => {
+                let args_bytes =
+                    hex::decode(args).map_err(|e| format!("Invalid args hex encoding: {}", e))?;
+
+                let _: () = ic_cdk::call(receiver_id, method_name.as_str(), (args_bytes,))
```

Comment:
ah, I thought `ic_call` would take the hex-encoded string, if it takes a `Vec<u8>` instead, then why are we hex-encoding?